### PR TITLE
Support non-file partials

### PR DESCRIPTION
**Pull request by pdbogen**
This PR does two things:

1. Supports non-file partials by abstracting partial retrieval to a PartialProvider class, and includes both a file-based implementation and map-based implementation; and
2. Integrates the new code into spec_test and enables tests that can now pass.

### CODE REVIEW

**Comment by cbroglie** [comment]
@pdbogen Thanks for the PR! I've been traveling and haven't had a chance to look yet, but will try to get to it soon

**Comment by pdbogen** [comment]
Sounds great. Please let me know if there are any issues, happy to fix them to get this merged.

File: mustache.go [hunk 1]
```diff
@@ -565,7 +548,9 @@ func renderElement(element interface{}, contextChain []interface{}, buf io.Write
 			return err
 		}
 	case *partialElement:
-		if err := elem.tmpl.renderTemplate(contextChain, buf); err != nil {
+		if partial, err := elem.prov.Get(elem.name); err != nil {
+			return err
+		} else if err := partial.renderTemplate(contextChain, buf); err != nil {
```
Comment (cbroglie) [inline mustache.go]: I find using the scoped variable in an else if kind of hard to read. Would prefer if this were written as:
```go
partial, err := elem.prov.Get(elem.name);
if err != nil {
	return err
}
if err := partial.renderTemplate(contextChain, buf); err != nil {
 	return err
}
```

**Comment by pdbogen** [comment]
Hi, @cbroglie. Sorry, in turn, for the delay on this.

I fixed the issues you pointed out. Let me add some commentary around the tags / compareTags test.. I've re-added it, but it's much simpler. We can't figure out what tags are contained within a Partial at any time other than render time; which means we can't do it just given a template. This is because Partials _can_ be recursive, and it's considered a runtime error if they recurse infinitely.

In order to determine the tags contained in a _template_ we would have to explore every possible branch (where "branch" is the part of mustache that says "if X is defined, render this ... "); which could create infinite loops in situations that would be acceptable/allowable (because the loop would terminate in a sane runtime rendering).

Hope that all makes sense...

Please let me know if you have any other comments or requests.

Thanks,
- Patrick

**Comment by cbroglie** [comment]
That makes sense, thanks again for the contribution.

**Pull request merged**